tests/unit/test_zarrio.py

```diff
@@ -182,3 +185,47 @@ def test_force_open_without_consolidated_fails(self):
             except ValueError as e:
                 self.fail("ZarrIO.__open_file_consolidated raised an unexpected ValueError: {}".format(e))
 
+
+class TestDatasetofReferences(ZarrStoreTestCase):
+    def setUp(self):
+        self.store_path = "test_io.zarr"
+        self.store = DirectoryStore(self.store_path)
+
+    def tearDown(self):
+        """
+        Remove all files and folders defined by self.store_path
+        """
+        paths = self.store_path if isinstance(self.store_path, list) else [self.store_path, ]
+        for path in paths:
+            if os.path.exists(path):
+                if os.path.isdir(path):
+                    shutil.rmtree(path)
+                elif os.path.isfile(path):
+                    os.remove(path)
+                else:
+                    warnings.warn("Could not remove: %s" % path)
+
+    def test_append_references(self):
+        # Setup a file container with references
+        num_bazs = 10
+        bazs = []  # set up dataset of references
+        for i in range(num_bazs):
+            bazs.append(Baz(name='baz%d' % i))
+        baz_data = BazData(name='baz_data', data=bazs)
+        container = BazBucket(bazs=bazs, baz_data=baz_data)
+        manager = get_baz_buildmanager()
+        # write to file
+        with ZarrIO(self.store, manager=manager, mode='w') as writer:
+            writer.write(container=container)
+        # read from file and validate references
+        with ZarrIO(self.store, manager=manager, mode='r+') as append_io:
+            read_container = append_io.read()
+            DoR = read_container.baz_data.data
+            DoR.append(DoR[0])
+
+            expected =  {'source': '.', 'path': '/bazs/baz0',
+                         'object_id': DoR[0].object_id,
+                         'source_object_id': container.object_id}
+
+            self.assertEqual(len(read_container.baz_data.data), 11)
+            self.assertDictEqual(read_container.baz_data.data.dataset[10], expected)
```

Comment:
Just in case, could you test closing this file, opening it, and making sure the size is still 11?